File: DAMOS/branches/0.3.0_RELEASE/tests/org.eclipselabs.damos.rte.posix.codegen.c.tests/src/org/eclipselabs/damos/rte/posix/codegen/c/gtests/CounterTestWithPrefix.java
/****************************************************************************
 * Copyright (c) 2008, 2012 Andreas Unger and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Andreas Unger - initial API and implementation 
 ****************************************************************************/

package org.eclipselabs.damos.rte.posix.codegen.c.gtests;

import org.eclipselabs.damos.codegen.c.test.GTest;
import org.eclipselabs.damos.dconfig.DconfigFactory;
import org.eclipselabs.damos.dconfig.PropertyDeclaration;
import org.eclipselabs.damos.dconfig.SimpleProperty;
import org.eclipselabs.damos.dconfig.SimplePropertyDeclaration;
import org.eclipselabs.damos.dconfig.util.PropertyEnumerationHelper;
import org.eclipselabs.damos.mscript.MscriptFactory;
import org.eclipselabs.damos.mscript.StringLiteral;

/**
 * Right now generated code is invalid if no prefix is specified.
 * This should be fixed in the future.
 * 
 * @author Andreas Unger
 */
@GTest(sourceFile="gtests/CounterTestWithPrefix.cpp", program="gtests/CounterTestWithPrefix")
public class CounterTestWithPrefix extends CounterTest {

	/* (non-Javadoc)
	 * @see org.eclipselabs.damos.codegen.c.test.AbstractGeneratorGTest#createConfiguration()
	 */
	@Override
	protected void createConfiguration() {
		super.createConfiguration();
		PropertyDeclaration prefixPropertyDeclaration = new PropertyEnumerationHelper().getPropertyDeclaration("damos.codegen.c.prefix");
		SimpleProperty prefixProperty = DconfigFactory.eINSTANCE.createSimpleProperty();
		prefixProperty.setDeclaration((SimplePropertyDeclaration) prefixPropertyDeclaration);
		StringLiteral prefix = MscriptFactory.eINSTANCE.createStringLiteral();
		prefix.setValue("CounterTest_");
		prefixProperty.setValue(prefix);
		prefixProperty.setPropagate(true);
		configuration.getRootSystemConfiguration().getBody().getProperties().add(prefixProperty);
	}
	
}

File: DAMOS/branches/initial-contribution/tests/org.eclipse.damos.rte.posix.codegen.c.tests/gtests/CounterTestWithPrefix.cpp
/*******************************************************************************
 * Copyright (c) 2008, 2012 Andreas Unger and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     Andreas Unger - initial API and implementation
 *******************************************************************************/
#include <unistd.h>
#include "gtest/gtest.h"
#include "CounterTestWithPrefix.h"

TEST(CounterTest, Counter) {
	CounterTest_Input input;
	CounterTest_Output output;

	CounterTest_initialize();

	for (int i = 0; i < COUNTERTEST_TASK_COUNT; ++i) {
		pthread_t thread;
		ASSERT_EQ(0, pthread_create(&thread, NULL, CounterTest_taskInfos[i].function, NULL));
	}

	input.a = 1;
	input.b = 0;
	CounterTest_execute(&input, &output);
	CounterTest_execute(&input, &output);
	CounterTest_execute(&input, &output);
	CounterTest_execute(&input, &output);
	CounterTest_execute(&input, &output);

	sleep(1);

	input.a = 0;
	input.b = 1;
	CounterTest_execute(&input, &output);
	CounterTest_execute(&input, &output);

	sleep(1);

	input.a = 0;
	input.b = 0;
	CounterTest_execute(&input, &output);

	EXPECT_EQ(6, output.out);
}
